File: train/twenty48/mcts.py
"""MCTS com chance nodes explícitos — espelho de play/src/core/mcts.ts, mas com
avaliação de folhas EM LOTE (crítico para eficiência de GPU).

Batching por ondas com virtual loss: cada onda desce `batch_size` vezes até
folhas distintas (o virtual loss diversifica as descidas e os chance nodes
injetam variação por si), avalia todas as folhas num único forward, e então
retropropaga. A semântica de busca é a mesma do TS (backup ponderado pela
distribuição exata de spawn via amostragem; valor somado sem alternância de
sinal, pois 2048 é solo contra o acaso).

O avaliador é a fronteira `evaluate` (evaluators.py): rollout no warm start,
rede no self-play. O núcleo abaixo não muda entre eles.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

from .board import (
    ACTIONS,
    GameState,
    apply_move,
    is_terminal,
    normalize_score,
    spawn_outcomes,
    with_spawn,
)

VLOSS_VALUE = 0.0  # 2048: valor em [0,1]; "perda" virtual = 0 (desencoraja reescolha)


class DecisionNode:
    __slots__ = (
        "state",
        "terminal",
        "expanded",
        "pending",
        "P",
        "N",
        "W",
        "VL",
        "legal",
        "children",
        "total_n",
        "total_vl",
        "eval_value",
    )

    def __init__(self, state: GameState):
        self.state = state
        self.terminal = is_terminal(state)
        self.expanded = False
        self.pending = False  # já enfileirado para avaliação nesta onda
        self.P = [0.0, 0.0, 0.0, 0.0]
        self.N = [0.0, 0.0, 0.0, 0.0]
        self.W = [0.0, 0.0, 0.0, 0.0]
        self.VL = [0, 0, 0, 0]  # virtual loss por aresta
        self.legal = [False, False, False, False]
        self.children: list[ChanceNode | None] = [None, None, None, None]
        self.total_n = 0
        self.total_vl = 0
        self.eval_value = 0.0


class ChanceNode:
    __slots__ = ("cells", "gained", "base_score", "size", "outcomes", "children")

    def __init__(self, cells, gained, base_score, size):
        self.cells = cells
        self.gained = gained
        self.base_score = base_score
        self.size = size
        self.outcomes = spawn_outcomes(cells)
        self.children: dict[int, DecisionNode] = {}


@dataclass
class MctsConfig:
    simulations: int = 128
    c_puct: float = 1.5
    batch_size: int = 32
    dirichlet_alpha: float = 0.5
    dirichlet_eps: float = 0.25


@dataclass
class SearchResult:
    visits: list[float]
    q_values: list[float]
    priors: list[float]
    legal: list[bool]
    best_action: int  # -1 se terminal


def _terminal_value(state: GameState) -> float:
    return normalize_score(state.score)


def _select(node: DecisionNode, c_puct: float) -> int:
    sqrt_total = math.sqrt(node.total_n + node.total_vl)
    best, best_a = -1.0, -1
    for a in ACTIONS:
        if not node.legal[a]:
            continue
        n_eff = node.N[a] + node.VL[a]
        q = node.W[a] / n_eff if n_eff > 0 else 0.0
        u = c_puct * node.P[a] * sqrt_total / (1 + n_eff)
        s = q + u
        if s > best:
            best, best_a = s, a
    return best_a


def _get_chance(node: DecisionNode, a: int) -> ChanceNode:
    existing = node.children[a]
    if existing is not None:
        return existing
    cells, gained, _ = apply_move(node.state, a)
    chance = ChanceNode(cells, gained, node.state.score, node.state.size)
    node.children[a] = chance
    return chance


def _sample_outcome(chance: ChanceNode, rng: np.random.Generator) -> DecisionNode:
    outcomes = chance.outcomes
    r = rng.random()
    chosen = outcomes[-1]
    for o in outcomes:
        r -= o.prob
        if r < 0:
            chosen = o
            break
    key = chosen.index * 4 + chosen.exponent
    cached = chance.children.get(key)
    if cached is not None:
        return cached
    child = DecisionNode(
        GameState(
            size=chance.size,
            cells=with_spawn(chance.cells, chosen.index, chosen.exponent),
            score=chance.base_score + chance.gained,
        )
    )
    chance.children[key] = child
    return child


def _descend(root: DecisionNode, c_puct: float, rng: np.random.Generator):
    """Desce aplicando virtual loss. Retorna (path, leaf|None, terminal_value|None)."""
    path: list[tuple[DecisionNode, int]] = []
    node = root
    while True:
        if node.terminal:
            return path, None, _terminal_value(node.state)
        if not node.expanded:
            return path, node, None
        a = _select(node, c_puct)
        node.VL[a] += 1
        node.total_vl += 1
        path.append((node, a))
        chance = _get_chance(node, a)
        node = _sample_outcome(chance, rng)


def _backup(path: list[tuple[DecisionNode, int]], value: float) -> None:
    for node, a in path:
        node.VL[a] -= 1
        node.total_vl -= 1
        node.N[a] += 1
        node.W[a] += value
        node.total_n += 1


def _expand(nodes: list[DecisionNode], evaluator) -> None:
    """Avalia folhas em lote e as expande (P mascarado/renormalizado por legalidade)."""
    states = [nd.state for nd in nodes]
    policies, values = evaluator(states)
    for i, nd in enumerate(nodes):
        legal_mask = [apply_move(nd.state, a)[2] for a in ACTIONS]
        p = policies[i]
        total = sum(p[a] for a in ACTIONS if legal_mask[a])
        for a in ACTIONS:
            nd.legal[a] = legal_mask[a]
            nd.P[a] = (p[a] / total) if (legal_mask[a] and total > 0) else 0.0
        nd.expanded = True
        nd.eval_value = float(values[i])


def _apply_dirichlet(root: DecisionNode, alpha: float, eps: float, rng: np.random.Generator) -> None:
    legal_idx = [a for a in ACTIONS if root.legal[a]]
    if len(legal_idx) < 2:
        return
    noise = rng.dirichlet([alpha] * len(legal_idx))
    for k, a in enumerate(legal_idx):
        root.P[a] = (1 - eps) * root.P[a] + eps * float(noise[k])


def run_mcts(
    root_state: GameState,
    evaluator,
    rng: np.random.Generator,
    config: MctsConfig,
    add_noise: bool = False,
) -> tuple[SearchResult, DecisionNode]:
    root = DecisionNode(root_state)
    if root.terminal:
        return SearchResult([0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [False] * 4, -1), root

    _expand([root], evaluator)
    if add_noise:
        _apply_dirichlet(root, config.dirichlet_alpha, config.dirichlet_eps, rng)

    done = 0
    while done < config.simulations:
        wave = min(config.batch_size, config.simulations - done)
        collected: list[tuple[list, DecisionNode | None, float | None]] = []
        leaves: list[DecisionNode] = []
        for _ in range(wave):
            path, leaf, tval = _descend(root, config.c_puct, rng)
            if leaf is not None and not leaf.pending:
                leaf.pending = True
                leaves.append(leaf)
            collected.append((path, leaf, tval))
            done += 1

        if leaves:
            _expand(leaves, evaluator)

        for path, leaf, tval in collected:
            value = tval if tval is not None else leaf.eval_value  # type: ignore[union-attr]
            _backup(path, value)
        for leaf in leaves:
            leaf.pending = False

    visits = list(root.N)
    q_values = [(root.W[a] / root.N[a]) if root.N[a] > 0 else 0.0 for a in ACTIONS]
    priors = list(root.P)
    best_action, best_visits = -1, -1.0
    for a in ACTIONS:
        if root.legal[a] and root.N[a] > best_visits:
            best_visits, best_action = root.N[a], a
    return SearchResult(visits, q_values, priors, list(root.legal), best_action), root


def select_move(visits: list[float], temperature: float, rng: np.random.Generator) -> int:
    """Escolhe a ação a partir das visitas: τ→0 = argmax, τ=1 = ∝ visitas."""
    v = np.asarray(visits, dtype=np.float64)
    if v.sum() == 0:
        return -1
    if temperature <= 1e-6:
        return int(v.argmax())
    logits = np.log(np.maximum(v, 1e-12)) / temperature
    probs = np.exp(logits - logits.max())
    probs /= probs.sum()
    return int(rng.choice(len(v), p=probs))

File: play/src/core/mcts.ts
// ============================================================================
// MCTS com CHANCE NODES EXPLÍCITOS (estilo AlphaZero, single-player estocástico)
// ----------------------------------------------------------------------------
// Alterna nós de decisão (4 ações, seleção PUCT) e nós de chance (spawn). O
// backup é ponderado pela distribuição EXATA de spawn: como os resultados são
// amostrados na descida ∝ sua probabilidade real, a média de visitas de um
// chance node é um estimador não-viesado da expectativa exata Σ p_i · V(i).
//
// Como 2048 é solo contra o acaso (sem oponente), o valor tem um único
// referencial "quão bom é o estado" — o backup soma o valor da folha em todas as
// arestas do caminho, sem alternância de sinal.
//
// A avaliação de folha entra SÓ pela interface Evaluator (fronteira neutra): o
// núcleo abaixo é idêntico com rollout stub (Fase 1) ou rede (Fase 3/4).
// ============================================================================

import type { Action, Evaluator, GameState } from './types';
import { ACTIONS } from './types';
import { type RNG } from './rng';
import { applyMove, isTerminal, spawnOutcomes, withSpawn, type SpawnOutcome } from './board';
import { normalizeScore } from './evaluate';

// ----------------------------------------------------------------------------
// Nós da árvore
// ----------------------------------------------------------------------------

interface DecisionNode {
  readonly state: GameState;
  readonly terminal: boolean;
  expanded: boolean;
  /** Prior de política por ação (do Evaluator). */
  readonly P: Float64Array;
  /** Visitas por aresta de ação. */
  readonly N: Float64Array;
  /** Soma de valores retropropagados por aresta. */
  readonly W: Float64Array;
  /** Legalidade por ação (definida na expansão). */
  readonly legal: boolean[];
  /** Chance node por ação (lazy). */
  readonly children: (ChanceNode | null)[];
  /** Soma das visitas das arestas (= Σ N). */
  totalN: number;
}

interface ChanceNode {
  /** Board pós-movimento, antes do spawn. */
  readonly cells: Uint8Array;
  /** Pontos ganhos no movimento que criou este nó. */
  readonly gained: number;
  /** Score do estado de decisão pai (para compor o score dos filhos). */
  readonly baseScore: number;
  readonly size: number;
  /** Distribuição exata de spawn a partir de `cells`. */
  readonly outcomes: SpawnOutcome[];
  /** Nós de decisão filhos, por resultado de spawn (lazy). */
  readonly children: Map<number, DecisionNode>;
}

function createDecision(state: GameState): DecisionNode {
  return {
    state,
    terminal: isTerminal(state),
    expanded: false,
    P: new Float64Array(4),
    N: new Float64Array(4),
    W: new Float64Array(4),
    legal: [false, false, false, false],
    children: [null, null, null, null],
    totalN: 0,
  };
}

// ----------------------------------------------------------------------------
// Configuração e resultado
// ----------------------------------------------------------------------------

export interface MctsConfig {
  /** Simulações por busca. */
  readonly simulations: number;
  /** Coeficiente de exploração do PUCT. */
  readonly cPuct: number;
  /** Avaliador de folha (stub na Fase 1, rede na Fase 3/4). */
  readonly evaluator: Evaluator;
  /** RNG para amostragem dos chance nodes. */
  readonly rng: RNG;
  /**
   * Valor de um estado terminal. Padrão: normalizeScore (o valor verdadeiro de
   * um fim de jogo é seu score final normalizado — quantidade conhecida, não
   * aprendida).
   */
  readonly terminalValue?: (state: GameState) => number;
}

export interface SearchResult {
  /** Contagem de visitas por ação — a "política do MCTS". */
  readonly visits: number[];
  /** Q médio por ação (0 se não visitada). */
  readonly qValues: number[];
  /** Prior de política da raiz (do Evaluator). */
  readonly priors: number[];
  /** Ações legais na raiz. */
  readonly legal: boolean[];
  /** Ação mais visitada entre as legais (-1 se terminal). */
  readonly bestAction: Action | -1;
  /** Total de simulações efetivamente rodadas. */
  readonly simulations: number;
}

// ----------------------------------------------------------------------------
// Núcleo da busca
// ----------------------------------------------------------------------------

function expand(node: DecisionNode, evaluator: Evaluator): number {
  const evaluation = evaluator(node.state);
  const policy = evaluation.policy;
  for (const a of ACTIONS) {
    node.P[a] = policy[a] ?? 0;
    node.legal[a] = applyMove(node.state, a).moved;
  }
  node.expanded = true;
  return evaluation.value;
}

function selectAction(node: DecisionNode, cPuct: number): Action {
  const sqrtTotal = Math.sqrt(node.totalN);
  let best = -Infinity;
  let bestAction: Action = 0;
  for (const a of ACTIONS) {
    if (!node.legal[a]) continue;
    const n = node.N[a];
    const q = n > 0 ? node.W[a] / n : 0;
    const u = (cPuct * node.P[a] * sqrtTotal) / (1 + n);
    const s = q + u;
    if (s > best) {
      best = s;
      bestAction = a;
    }
  }
  return bestAction;
}

/** Chance node para a ação `a` (assume-se legal); criado sob demanda. */
function getChance(node: DecisionNode, a: Action): ChanceNode {
  const existing = node.children[a];
  if (existing) return existing;
  const res = applyMove(node.state, a);
  const chance: ChanceNode = {
    cells: res.cells,
    gained: res.gained,
    baseScore: node.state.score,
    size: node.state.size,
    outcomes: spawnOutcomes(res.cells),
    children: new Map(),
  };
  node.children[a] = chance;
  return chance;
}

/** Amostra um resultado de spawn ∝ probabilidade exata e desce para o filho. */
function sampleOutcome(chance: ChanceNode, rng: RNG): DecisionNode {
  const outcomes = chance.outcomes;
  let r = rng.next();
  let chosen = outcomes[outcomes.length - 1];
  for (const o of outcomes) {
    r -= o.prob;
    if (r < 0) {
      chosen = o;
      break;
    }
  }
  const key = chosen.index * 4 + chosen.exponent;
  const cached = chance.children.get(key);
  if (cached) return cached;
  const child = createDecision({
    size: chance.size,
    cells: withSpawn(chance.cells, chosen.index, chosen.exponent),
    score: chance.baseScore + chance.gained,
  });
  chance.children.set(key, child);
  return child;
}

interface PathStep {
  node: DecisionNode;
  action: Action;
}

function simulate(
  root: DecisionNode,
  evaluator: Evaluator,
  cPuct: number,
  rng: RNG,
  terminalValue: (state: GameState) => number,
): void {
  const path: PathStep[] = [];
  let node = root;
  let value: number;

  for (;;) {
    if (node.terminal) {
      value = terminalValue(node.state);
      break;
    }
    if (!node.expanded) {
      value = expand(node, evaluator);
      break;
    }
    const action = selectAction(node, cPuct);
    path.push({ node, action });
    const chance = getChance(node, action);
    node = sampleOutcome(chance, rng);
  }

  for (const step of path) {
    step.node.N[step.action] += 1;
    step.node.W[step.action] += value;
    step.node.totalN += 1;
  }
}

/** Roda a busca a partir de `rootState` e retorna as estatísticas de visita. */
export function runMcts(rootState: GameState, config: MctsConfig): SearchResult {
  const terminalValue = config.terminalValue ?? ((s: GameState) => normalizeScore(s.score));
  const root = createDecision(rootState);

  if (root.terminal) {
    return {
      visits: [0, 0, 0, 0],
      qValues: [0, 0, 0, 0],
      priors: [0, 0, 0, 0],
      legal: [false, false, false, false],
      bestAction: -1,
      simulations: 0,
    };
  }

  expand(root, config.evaluator);
  for (let s = 0; s < config.simulations; s++) {
    simulate(root, config.evaluator, config.cPuct, config.rng, terminalValue);
  }

  const visits = Array.from(root.N);
  const qValues = ACTIONS.map((a) => (root.N[a] > 0 ? root.W[a] / root.N[a] : 0));
  const priors = Array.from(root.P);

  let bestAction: Action | -1 = -1;
  let bestVisits = -1;
  for (const a of ACTIONS) {
    if (root.legal[a] && root.N[a] > bestVisits) {
      bestVisits = root.N[a];
      bestAction = a;
    }
  }

  return {
    visits,
    qValues,
    priors,
    legal: [...root.legal],
    bestAction,
    simulations: config.simulations,
  };
}
